src/shared/switch-root.c

```diff
@@ -82,9 +81,9 @@ int switch_root(const char *new_root,
                 /* When the path was not a mount point, then we need to reopen the path, otherwise, it still
                  * points to the underlying directory. */
 
-                fd = open(new_root, O_DIRECTORY|O_CLOEXEC);
+                fd = fd_reopen(new_root_fd, O_DIRECTORY|O_CLOEXEC);
```

Comment:
The original problem was that the fd points to the underlying dir instead of the mountpoint. Therefore, I don't think this is correct?

Reply:
the kernel will on open() re-resolve an inode, and ensure we only look at the topmost mount.

Reply:
that's at least the theory

Reply:
hmm, now i wonder if this behaves differently in the kernel if you do `openat(fd, ".")` instead of `open("/proc/self/$fd")`.... i.e. we might need the former here, not the latter